src/audio/drc/drc_hifi4.c

```diff
@@ -0,0 +1,911 @@
+		 */
+		tmp = AE_SLAI32S(compression_diff_db, 3); /* Q11.21 -> Q8.24 */
+		state->max_attack_compression_diff_db =
+			AE_MAX32(state->max_attack_compression_diff_db, tmp);
+
+		eff_atten_diff_db =
+			MAX(HALF_Q24, state->max_attack_compression_diff_db); /* Q8.24 */
+
+		/* x = 0.25f / eff_atten_diff_db;
+		 * => x = 1.0f / (eff_atten_diff_db << 2);
+		 */
+		x = drc_inv_fixed(eff_atten_diff_db, 22 /* Q8.24 << 2 */, 26); /* Q6.26 */
+		envelope_rate = AE_SUB32(ONE_Q20,
+					 drc_pow_fixed(x, p->one_over_attack_frames)); /* Q12.20 */
+	}
+
+	tmp = AE_SLAI32S(envelope_rate, 10); /* Q12.20 -> Q2.30 */
+	state->envelope_rate = tmp;
+	state->scaled_desired_gain = scaled_desired_gain;
+}
+
+/* Calculate compress_gain from the envelope and apply total_gain to compress
+ * the next output division.
+ */
+void drc_compress_output(struct drc_state *state,
+			 const struct sof_drc_params *p,
+			 int nbyte,
+			 int nch)
+{
+	const int count = DRC_DIVISION_FRAMES >> 2;
+	ae_f32 x[4]; /* Q2.30 */
+	ae_f32 c, base, r, r2, r4; /* Q2.30 */
+	ae_f32 post_warp_compressor_gain;
+	ae_f32 total_gain;
+	int i, j, ch;
+	int inc = state->pre_delay_read_index;
+	int16_t *sample16_p; /* for s16 format case */
+	int32_t *sample32_p; /* for s24 and s32 format cases */
+	int32_t sample;
+	int is_2byte = (nbyte == 2); /* otherwise is 4-bytes */
+	ae_f32 tmp;
+	ae_f64 tmp64;
+	ae_f32 master_linear_gain = p->master_linear_gain;
+
+	/* Exponential approach to desired gain. */
+	if (state->envelope_rate < ONE_Q30) {
+		/* Attack - reduce gain to desired. */
+		c = AE_SUB32(state->compressor_gain, state->scaled_desired_gain);
+		base = state->scaled_desired_gain;
+		r = AE_SUB32(ONE_Q30, state->envelope_rate);
+		x[0] = drc_mult_lshift(c, r, LSHIFT_QX30_QY30_QZ30);
+		for (j = 1; j < 4; j++)
+			x[j] = drc_mult_lshift(x[j - 1], r, LSHIFT_QX30_QY30_QZ30);
+		r2 = drc_mult_lshift(r, r, LSHIFT_QX30_QY30_QZ30);
+		r4 = drc_mult_lshift(r2, r2, LSHIFT_QX30_QY30_QZ30);
+
+		i = 0;
+		if (is_2byte) { /* 2 bytes per sample */
+			while (1) {
+				for (j = 0; j < 4; j++) {
+					/* Warp pre-compression gain to smooth out sharp
+					 * exponential transition points.
+					 */
+					tmp = AE_ADD32(x[j], base);
+					post_warp_compressor_gain = drc_sin_fixed(tmp); /* Q1.31 */
+
+					/* Calculate total gain using master gain. */
+					tmp64 = AE_MULF32R_LL(master_linear_gain,
+							      post_warp_compressor_gain);
+					total_gain = AE_ROUND32F48SSYM(tmp64); /* Q8.24 */
+
+					/* Apply final gain. */
+					for (ch = 0; ch < nch; ch++) {
+						sample16_p =
+							(int16_t *)state->pre_delay_buffers[ch] +
+							inc;
+						sample = (int32_t)*sample16_p;
+						sample = drc_mult_lshift(sample, total_gain,
+									 LSHIFT_QX15_QY24_QZ15);
+						*sample16_p = sat_int16(sample);
+					}
+					inc++;
+				}
+
+				if (++i == count)
+					break;
+
+				for (j = 0; j < 4; j++)
+					x[j] = drc_mult_lshift(x[j], r4, LSHIFT_QX30_QY30_QZ30);
+			}
```

Comment:
alternatively this could be written as
```
for (i = 0; i < count; i++) {
    for (j = 0; j < 4; j++) {
        ...
        if (i < count - 1)
            x[j] = drc_mult_lshift(x[j], r4, LSHIFT_QX30_QY30_QZ30);
    }
}
```
but I'm not sure if it wins anything in terms of execution time, at least instead of 2 loops and an `if` internally it then would have 1 loop and an `if`. Same 3 more times below of course

Reply:
I tried before and checked again, wouldn't better, so keep the original version

Reply:
it is not be better because j < 4, if it is j < 48 or larger, it would bring better cache performance.
this is due to origin code need D cache load more more data than proposed.

I agree keep as current, what's 4 means here?